### PR TITLE
Add ability to configure kubeadm's ignored pre-flight errors via InitConfiguration and JoinConfiguration

### LINKED ISSUE #1460: Add way to specify kubeadm's --ignore-preflight-errors in InitConfiguration and JoinConfiguration
**What would you like to be added**:

A single way to configure all of `kubeadm`'s behaviour.
In particular, given `ignorePreflightErrors` and `externalcfg` are currently disjoint:

- https://github.com/kubernetes/kubernetes/blob/a887ae8344adf9739c1d156a718c05fdf95b73ab/cmd/kubeadm/app/cmd/init.go#L104-L106
- https://github.com/kubernetes/kubernetes/blob/a887ae8344adf9739c1d156a718c05fdf95b73ab/cmd/kubeadm/app/cmd/join.go#L132-L133

I would like a way to specify the set of strings corresponding to `--ignore-preflight-errors`'s value as an "array" in (at least) `InitConfiguration` and `JoinConfiguration`.

**Why is this needed**:

This would simplify programmatic usage of `kubeadm`: one configuration file to generate, and then pass to `kubeadm`.

**Potential way forward**:

1. Keep `--ignore-preflight-errors` for backward-compatibility (for now).
2. Add the following field to both `InitConfiguration` and `JoinConfiguration`:

    ```go
    IgnorePreflightErrors []string `json:"ignorePreflightErrors,omitempty"`
    ```

3. Take the union of both sets, validate it (as currently done), and use this going in the rest of the program.

**Misc.**:

Happy to make the change myself, once I get a "go" (or at least, no "no-go") from someone more knowledgeable. (Maybe @luxas? 🙂 )

/kind feature
/sig cluster-lifecycle

### LINKED ISSUE #74246: Add way to specify kubeadm's --ignore-preflight-errors in InitConfiguration and JoinConfiguration
**What would you like to be added**:

A single way to configure all of `kubeadm`'s behaviour.
In particular, given `ignorePreflightErrors` and `externalcfg` are currently disjoint:
https://github.com/kubernetes/kubernetes/blob/a887ae8344adf9739c1d156a718c05fdf95b73ab/cmd/kubeadm/app/cmd/init.go#L104-L106
https://github.com/kubernetes/kubernetes/blob/a887ae8344adf9739c1d156a718c05fdf95b73ab/cmd/kubeadm/app/cmd/join.go#L132-L133
I would like a way to specify the set of strings corresponding to `--ignore-preflight-errors`'s value as an "array" in (at least) `InitConfiguration` and `JoinConfiguration`.

**Why is this needed**:

This would simplify programmatic usage of `kubeadm`: one configuration file to generate, and then pass to `kubeadm`.

**Potential way forward**:

1. Keep `--ignore-preflight-errors` for backward-compatibility (for now).
2. Add the following field to both `InitConfiguration` and `JoinConfiguration`:

    ```go
    IgnorePreflightErrors []string `json:"ignorePreflightErrors,omitempty"`
    ```

3. Take the union of both sets, validate it (as currently done), and use this going in the rest of the program.

**Misc.**:

Happy to make the change myself, once I get a "go" (or at least, no "no-go") from someone more knowledgeable. (Maybe @luxas? 🙂 )

### PR DESCRIPTION
**Pull request by marccarre**
**What type of PR is this?**
/kind feature
/sig cluster-lifecycle

**What this PR does / why we need it**:
This adds a more declarative way to configure ignored pre-flight errors -- via `InitConfiguration` and `JoinConfiguration`, in addition to the "legacy" CLI flag -- and therefore simplifies programmatic usage of `kubeadm`: just one configuration file to generate & to pass to `kubeadm`.

**Which issue(s) this PR fixes**:
Fixes #74246.
Fixes kubernetes/kubeadm/issues/1460.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm's ignored pre-flight errors can now be configured via InitConfiguration and JoinConfiguration.
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: this is already planned in the v1beta2 config, see:
https://github.com/kubernetes/kubeadm/issues/1439

so technically we can accept this PR but only once the v1beta1 types are already created, and we cannot add this to v1beta1. today we are going to have a kubeadm 1.15 planning session at 4PM UTC.
https://github.com/kubernetes/community/tree/master/sig-cluster-lifecycle

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: i think we had a similar utility function in another location.

**Review by neolit123 — COMMENTED** [review]
thanks for the PR @marccarre 
i think, this is pretty much what we want to do, except that it has to happen in v1beta2.

/assign @rosti @fabriziopandini 
/hold

**Comment by neolit123** [comment]
@marccarre 
please squash the commits into one.
/priority important-longterm

**Review by rosti — COMMENTED** [review]
Thanks for your contribution @marccarre !
The thing is that we should not modify the existing v1beta1 config. We have to do that modification in the new v1beta2 config if we reach a consensus on that.
Also, my proposition is to have that field in the `NodeRegistrationOptions` only.

Let's hold this for now, until the time comes for it.
/hold

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: I'd be interested in it if you find it again. `git grep`-ing for `func Union`/`func union` didn't yield anything.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: TIL, thanks for pointing to this issue.
I have undone the changes on v1beta1, but now see the below error:
```console
$ make WHAT=cmd/kubeadm
[...]
E0321 15:15:07.858860   27590 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm.NodeRegistrationOptions -> k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha3.NodeRegistrationOptions
E0321 15:15:07.858900   27590 conversion.go:668]   the following fields need manual conversion:
E0321 15:15:07.858910   27590 conversion.go:670]       - IgnorePreflightErrors
E0321 15:15:07.972950   27590 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm.NodeRegistrationOptions -> k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1.NodeRegistrationOptions
E0321 15:15:07.972989   27590 conversion.go:668]   the following fields need manual conversion:
E0321 15:15:07.972996   27590 conversion.go:670]       - IgnorePreflightErrors
[...]
make: *** [all] Error 1
```
...and it isn't obvious where this manual conversion should be added. 
Any pointer?

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: Alternatively, I could move this to some utility file. Any preference on the location? (I was thinking of moving it under `cmd/kubeadm/app/util/sets.go`)

**Comment by marccarre** [comment]
@rosti, @neolit123, thanks for your feedback ✨ 
I've addressed some of it (and will squash at the very end once everyone is happy about this PR), but now face build issues as per https://github.com/kubernetes/kubernetes/pull/75499#discussion_r267634175. Apologies for being new to `kubeadm`/`kubernetes`' development, but any pointer on how to resolve this would be very much appreciated.

File: cmd/kubeadm/app/cmd/init.go
Comment (kad) [reply cmd/kubeadm/app/cmd/init.go]: you can do in the code above something like this:
```go

ignorePreflightErrors := sets.NewString(options.ignorePreflightErrors).Union(sets.NewString(cfg.NodeRegistration.IgnorePreflightErrors)).List()
```

**Review by rosti — COMMENTED** [review]
Thanks for updating this @marccarre !
However, we need to hold on until the new config (probably v1beta2) becomes available. More developments on this question are to come probably next week.
We need to add the new field in that config and the internal one (not in v1beta1 or v1alpha3).
I'll ping you for a rebase and more guidance once we are ready for this.
It's just that we need to do a few more things before we can proceed with this.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Ha! I had removed them initially when experimenting, but eventually added them back as I thought they didn't hurt. I've just removed them again. Thanks again for the feedback!

**Comment by marccarre** [comment]
> we need to hold on until the new config (probably v1beta2) becomes available

Sure, no worries. I'll wait then 🙂 

I haven't dug into the code generation logic (yet), but I'm guessing some of the build issues I'm seeing will change/disappear once `v1beta2` is added.

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: Yes, I thought of this too, but found it relatively verbose, so thought I'd hide the details behind a function. I can do this if there is a strong opinion though. Thanks @kad for pointing this out in any case 👍

**Comment by rosti** [comment]
> I haven't dug into the code generation logic (yet), but I'm guessing some of the build issues I'm seeing will change/disappear once `v1beta2` is added.

Mostly yes.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: The tag will be needed in the versioned config (v1beta2) though. But since that is non-existent as of today, there is not much you can do ATM.

**Comment by rosti** [comment]
@marccarre, with the merger of #76710 you are now able to continue working on this based on v1beta2.
In addition to that we need to verify the new field and make sure it cannot be set to "all" and therefore ignore all pre-flight errors from the config easily.
Thanks for taking this on!

**Comment by kad** [comment]
I thought about this PR, and I think the better way to implement that will be to modify `validation.ValidateIgnorePreflightErrors` such, that it will take two arrays of strings as arguments. One from command line, one from config objects. Then, merging into one set can be done inside `validation.ValidateIgnorePreflightErrors` by iterating both of the arrays, practically getting rid of helper function `union`. Also it would be good to validate that array from config object will not contain 'all', as config object should not be used as "hide everything" workarround, instead of explicitly specifying what should be ignored.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: Resolved now that I've rebased on top of latest `master`.

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: Resolved now that the union is done directly by `ValidateIgnorePreflightErrors` as suggested in https://github.com/kubernetes/kubernetes/pull/75499#issuecomment-492216224.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Resolved: I had removed the tag on `cmd/kubeadm/app/apis/kubeadm/types.go` a while ago, and I've now added the field _with_ the tag in the `v1beta2` file.

**Comment by marccarre** [comment]
@rosti, @fabriziopandini, @neolit123, @kad, I've just rebased this PR on top of latest `master`, added the necessary changes for `v1beta2` and addressed the feedback given so far (thank you all! 🙇). PTAL 🙂

**Comment by kad** [comment]
Please squash commits. One for implementation/code changes part, one commit for auto-generated stuff.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go]: This needs a fairly simple test case.

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go]: It's worth noting here that, although "all" is allowed on the command line, it's not allowed in the config.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Ok, this is a little bit different, from what I expected.
The thing is that we want "all" to be an invalid value for errors specified in the config YAML.
Hence, it's beset to keep more or less the previous variant of the func, add a bool parameter to it (to tell if the error list is from the config file or the command line) and then denying "all" if the bool says it's from the config file.

File: cmd/kubeadm/app/cmd/init.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/init.go]: Ideally we should call this twice - once for command line supplied errors and once for config file supplied ones. This is because of the "all" keyword (it must not be specified in the config file).
After that, we should have a couple of sets, that need merging. If the command line set has "all" in it, the resulting set should be consisting only of `{"all"}`. Otherwise it's a simple merge of both. Probably a helper func should do the trick.

File: cmd/kubeadm/app/cmd/init_test.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/init_test.go]: You can have the `expected` parameter as a string slice and then convert it to a set with `NewString`. This will make the calls less verbose by omitting the extra `NewString` call in them.

**Review by rosti — COMMENTED** [review]
Thanks for the update @marccarre !
So far it looks great!
However, we need to deny the ability to specify "all" in the config file (but still permit it in the command line). Once that is done we can merge it.

**Comment by rosti** [comment]
> I thought about this PR, and I think the better way to implement that will be to modify `validation.ValidateIgnorePreflightErrors` such, that it will take two arrays of strings as arguments. One from command line, one from config objects. Then, merging into one set can be done inside `validation.ValidateIgnorePreflightErrors` by iterating both of the arrays, practically getting rid of helper function `union`. Also it would be good to validate that array from config object will not contain 'all', as config object should not be used as "hide everything" workarround, instead of explicitly specifying what should be ignored.

That's one way of doing it. 
I would like to keep things simple and scope of functionality clear, hence `ValidateIgnorePreflightErrors` should still perform only validation of errors, but with a bool param stating the origin of the ignored errors. This is where "all" from a config file should be checked.
Then a wrapper func, that takes a couple of sets (one from command line, another from the config) would call `ValidateIgnorePreflightErrors` twice. If the "all" keyword is present in the command line errors, it's returned as a set with a single element. Otherwise, the sets need to be combined.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (kad) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: @rosti, my suggestion was instead of `...[]string` to have two explicit arguments - one from command line, one from config. then to have two for loops, one processing config, then checking if 'all' is present in set (error if yes), then add overrides from command line. No need for bool here.

File: cmd/kubeadm/app/cmd/init.go
Comment (kad) [reply cmd/kubeadm/app/cmd/init.go]: @rosti  if config has something specific, and command line says `ignore all` we still need to fail. We shouldn't allow to mix 'all' with anything else. Let admin think twice (or more times) when admin tries to mask all errors.

**Comment by kad** [comment]
> Then a wrapper func, that takes a couple of sets (one from command line, another from the config) would call `ValidateIgnorePreflightErrors` twice. If the "all" keyword is present in the command line errors, it's returned as a set with a single element. Otherwise, the sets need to be combined.

This will require third iteration over the combined set with check, for mixture or `all` plus individual check, like it is currently performed in `validation.ValidateIgnorePreflightErrors`.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go]: Done. Will push soon.

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go]: Thanks for clarifying this. I indeed did not know this requirement.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Sure, I'll go for this then.

File: cmd/kubeadm/app/cmd/init_test.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init_test.go]: Indeed better, well spotted 👍 
Will push soonish.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Note that if we go for two explicit arguments (1), whether at this level, or in a wrapper function, maintainers would have to ensure arguments are passed in the right order.

If we go for a boolean (2), is may be a tiny bit more explicit, but the boolean would still be rather obscure, i.e. one would have to look at its name in the implementation to know its meaning.

A third approach (3), more verbose, but more explicit, would be to have two methods, e.g.`ValidateIgnorePreflightErrorsFromCLI` and `ValidateIgnorePreflightErrorsFromConfigFile` (or so, better names always welcome).

@kad, @rosti, please let me know your preferences.
(At the moment, I'm hesitating between (1) and (3).)

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: This indeed sounds like a desirable behaviour to me.

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go]: Solved, I think. Will push soon.

**Comment by marccarre** [comment]
I hope to have addressed both your requirements.
PTAL, @rosti, @kad, and thanks again for your feedback! ✨ 
(Previous push had some tests failing, if they still are after this push, I'll then look into this.)

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go
Comment (kad) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: not correct expected result. should be error. 'all' should not be mixed with anything else.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (kad) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: If 'all' is set in CLI and config has something else in addition, it is an error situation. 
```go
func ValidateIgnorePreflightErrors(ignorePreflightErrorsFromCLI, ignorePreflightErrorsFromConfigFile []string) (sets.String, error) {
	ignoreErrors := sets.NewString()
	allErrs := field.ErrorList{}

	for _, item := range ignorePreflightErrorsFromConfigFile {
		ignoreErrors.Insert(strings.ToLower(item)) // parameters are case insensitive
	}

	if ignoreErrors.Has("all") {
		// "all" is forbidden in config files. Administrators should use an
		// explicit list of errors they want to ignore, as it can be risky to
		// mask all errors in such a way. Hence, we return an error:
		allErrs = append(allErrs, field.Invalid(field.NewPath("ignorePreflightErrors"), "all", "'all' cannot be used in configuration file"))
        }

	for _, item = range ignorePreflightErrorsFromCLI {
		ignoreErrors.Insert(strings.ToLower(item)) // parameters are case insensitive
	}

	if ignoreErrors.Has("all") && ignoreErrors.Len() > 1 {
		allErrs = append(allErrs, field.Invalid(field.NewPath("ignore-preflight-errors"), strings.Join(ignoreErrors.List(), ","), "don't specify individual checks if 'all' is used"))
	}

	return ignoreErrors, allErrs.ToAggregate()
}
```

**Review by kad — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: My understanding was that **if** `all` is provided via the CLI, it should override the config file.
Is that not what we'd like?

Comment (kad) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: This forces to limit usage of 'all' as much as possible. Easier to add any individual check, rather than trying to blindly ignoring all possible errors from checks without thinking. We seen a lot of 'guides' in the internet where they all started with 'kubeadm .... --skip-preflight-checks' and then trying to deal with problems that were actually reported by preflight checks as errors/warnings.
So, morale for admins: if you really know what you're doing, think again, and if still think that error must be ignored - add individual error to the list. If you copy&pasting from somewhere 'all', think 3 times, are you really want to do that.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: (1) currently implemented.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (kad) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Small nit, not critical, but if you'll be changing due to failed tests, better to have only one `return` statement, as in previous version of function, and return all possible errors at once, instead of making admin to fix one error condition and then hit second one.

File: cmd/kubeadm/app/cmd/init.go
Comment (kad) [inline cmd/kubeadm/app/cmd/init.go]: Also nit: do we have usage of this somewhere? Otherwise we are loosing prestine state as we read from config file, isn't it?

**Review by kad — COMMENTED** [review]
small nits

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: As @kad mentioned below, let's drop this return and the one bellow and return `allErrs.ToAggregate()` in the last return at the end of the func.

**Review by rosti — COMMENTED** [review]
Thanks @marccarre !
A couple of nits and we are there!

**Comment by rosti** [comment]
Ok, the fuzzer test is failing, because we don't have the `IgnorePreflightErrors` in v1beta1. To fix this, we must reintroduce the `fuzzNodeRegistration` in the fuzzer test and plug the `IgnorePreflightErrors` field.
#75261 is the PR that got it deleted. You can check with it how to reintroduce it.
The new version of it should roughly look like this:
```golang
func fuzzNodeRegistration(obj *kubeadm.NodeRegistrationOptions, c fuzz.Continue) {	
	c.FuzzNoCustom(obj)	

 	// Pinning values for fields that get defaults if fuzz value is empty string or nil (thus making the round trip test fail)	
	obj.IgnorePreflightErrors = []string{}
}
```

Also, the reset test needs fixing (missing parameter in the call to `ValidateIgnorePreflightErrors`).

File: cmd/kubeadm/app/cmd/init.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/init.go]: This is currently not used anywhere, but given the configuration object is passed around, and this field _could_ eventually be used elsewhere as a result, I thought doing the above might reduce the chance of a bug/inconsistent behaviour. I can remove it if you'd rather keep the configuration file's initial state. Let me know.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Resolved in latest version. Thanks for the feedback 👍

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (marccarre) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Resolved in latest version. Thanks for the feedback 👍

File: cmd/kubeadm/app/cmd/join_test.go
Comment (marccarre) [inline cmd/kubeadm/app/cmd/join_test.go]: It seems importing `"k8s.io/apimachinery/pkg/util/sets"` in tests yields to build errors like [that one](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/75499/pull-kubernetes-verify/1129715477703036929):
```diff
diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD /tmp/verify-bazel.STLhDZ/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD
--- /home/prow/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD	2019-05-18 11:50:23.049528531 +0000
+++ /tmp/verify-bazel.STLhDZ/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/BUILD	2019-05-18 11:51:00.467822797 +0000
@@ -99,6 +99,7 @@ go_test(
         "//staging/src/k8s.io/apimachinery/pkg/api/errors:go_default_library",
         "//staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
         "//staging/src/k8s.io/apimachinery/pkg/runtime:go_default_library",
+        "//staging/src/k8s.io/apimachinery/pkg/util/sets:go_default_library",
         "//staging/src/k8s.io/client-go/kubernetes/fake:go_default_library",
         "//staging/src/k8s.io/client-go/testing:go_default_library",
         "//staging/src/k8s.io/client-go/tools/clientcmd:go_default_library",

Run ./hack/update-bazel.sh
```
This presumably happens because Bazel (which I'm unfamiliar with) requires this dependency to be declared. I have added it manually in `cmd/kubeadm/app/cmd/BUILD` but is there a better way to achieve this?

Comment (rosti) [reply cmd/kubeadm/app/cmd/join_test.go]: Yes, there is a script - `./hack/update-bazel.sh` that will do everything for you. Actually, always use that script to update bazel deps.

**Review by rosti — APPROVED** [review]
Thank you @marccarre !
I am happy how things turned out. Can you squash your commits as a final step so we can merge it?
/approve
/hold
For the squash.

File: cmd/kubeadm/app/cmd/join_test.go
Comment (marccarre) [reply cmd/kubeadm/app/cmd/join_test.go]: Thanks for the explanation! 👍

**Comment by marccarre** [comment]
Done: PR rebased on top of latest `master` and commits squashed.
Thanks again for the feedback! 👍

**Comment by rosti** [comment]
@kad can you do a final pass and lgtm if you are happy with this?
/assign @kad 
/hold cancel
/test pull-kubernetes-e2e-gce

**Comment by kad** [comment]
/lgtm
I have some doubts about some corner cases like in order of operations in upgrade steps (e.g. validating flags before/after trying to load cluster config), but I think it is not a big deal at the moment.

**Pull request merged**
